### PR TITLE
Filter stakes by withdrawer

### LINKED ISSUE #25462: Add way from CLI to find all the stake accounts for a given withdraw authority
#### Problem

There is not a way from the CLI (that I know of) to find all the stake accounts for a given withdraw authority. Both [Step Finance](https://app.step.finance/) and [Sonar Watch](https://sonar.watch/) support this query, and the CLI should be able to as well.

#### Proposed Solution

From @mvines

> `solana stakes` _could_ easily do this too. Here’s the gist of a patch to do so, sans the necessary plumbing to surface a new cli option to the user:
> 
> ```diff
> diff --git a/cli/src/cluster_query.rs b/cli/src/cluster_query.rs
> index f39b64cde0..02004fb590 100644
> --- a/cli/src/cluster_query.rs
> +++ b/cli/src/cluster_query.rs
> @@ -1753,9 +1753,9 @@ pub fn process_show_stakes(
>                      ),
>                      encoding: Some(rpc_filter::MemcmpEncoding::Binary),
>                  }),
> -                // Filter by `Delegation::voter_pubkey`, which begins at byte offset 124
> +                // Filter by `Meta::authorized::withdrawer`, which begins at byte offset 44
>                  rpc_filter::RpcFilterType::Memcmp(rpc_filter::Memcmp {
> -                    offset: 124,
> +                    offset: 44,
>                      bytes: rpc_filter::MemcmpEncodedBytes::Base58(
>                          vote_account_pubkeys[0].to_string(),
>                      ),
> @@ -1795,11 +1795,11 @@ pub fn process_show_stakes(
>                          });
>                      }
>                  }
> -                StakeState::Stake(_, stake) => {
> +                StakeState::Stake(meta, _stake) => {
>                      if vote_account_pubkeys.is_none()
>                          || vote_account_pubkeys
>                              .unwrap()
> -                            .contains(&stake.delegation.voter_pubkey)
> +                            .contains(&meta.authorized.withdrawer)
>                      {
>                          stake_accounts.push(CliKeyedStakeState {
>                              stake_pubkey: stake_pubkey.to_string(),
> ```

### PR DESCRIPTION
**Pull request by godmodegalactus**
#### Problem

Currently we cannot search stakes and filtering by withdraw authority. I have added into cli stakes command a way to filter it with withdraw authority.

#### Summary of Changes

Updating cli to pass withdrawer_pubkey and filter the results out by doing memcmp on the stakes.

Fixes #25462

### CODE REVIEW

File: cli/src/cluster_query.rs
Comment (joncinque) [inline cli/src/cluster_query.rs:345]: nit, for consistency with others like "stake-authorize"
```suggestion
                    .value_name("PUBKEY")
```

File: cli/src/cluster_query.rs
Comment (joncinque) [inline cli/src/cluster_query.rs:1823]: You can get rid of this filter, since withdraw authority exists on `StakeState::Stake(_,_)` and `StakeState::Initialized(_)`

File: cli/src/cluster_query.rs
Comment (joncinque) [inline cli/src/cluster_query.rs:1831]: Can you avoid the copy-pasta on this struct?

File: cli/src/cluster_query.rs
Comment (joncinque) [inline cli/src/cluster_query.rs:1855]: No need for the condition here since `Initialized` stakes are ok

File: cli/src/cluster_query.rs
Comment (joncinque) [inline cli/src/cluster_query.rs:1874]: This will be covered by the RPC filter, so it can be removed as well.

**Review by joncinque — COMMENTED** [review]
Thanks for the contribution! This is definitely on the right track, just a few bits to iron out and it'll be good to go.

File: cli/src/cluster_query.rs
Comment (CriesofCarrots) [inline cli/src/cluster_query.rs:333]: I'd prefer to keep this to be explicit about not breaking current users.

File: cli/src/cluster_query.rs
Comment (CriesofCarrots) [inline cli/src/cluster_query.rs:1802]: Is this comment on the wrong line?

File: cli/src/cluster_query.rs
Comment (CriesofCarrots) [reply cli/src/cluster_query.rs:347]: I think the `pubkey!` macro needs a period and a space at the end of this string, or it looks crummy

Comment (brooksprumo) [reply cli/src/cluster_query.rs:347]: How about "Only show stake accounts with the provided withdraw authority. " so it's similar to `vote_account_pubkeys`?

File: cli/src/cluster_query.rs
Comment (godmodegalactus) [reply cli/src/cluster_query.rs:1874]: Yes i agree but it was done for vote_account_pubkeys which seems weird as it also be covered by RPC filter.

Comment (godmodegalactus) [reply cli/src/cluster_query.rs:1874]: Ahh yes I understand.

File: cli/src/cluster_query.rs
Comment (brooksprumo) [inline cli/src/cluster_query.rs:348]: How about "Only show stake accounts with the provided withdraw authority. "?

https://github.com/solana-labs/solana/pull/25939#discussion_r897211030

File: cli/src/cluster_query.rs
Comment (brooksprumo) [inline cli/src/cluster_query.rs:345]: nit: Maybe this is just me, but why not call this arg `withdraw_authority` or `withdraw_authority_pubkey`? (and same for the variables that use this value)

File: cli/src/cluster_query.rs
Comment (godmodegalactus) [reply cli/src/cluster_query.rs:348]: I think "Filter" is more technical than "Only show stake", but if more reviewers prefer later why not.

File: cli/src/cluster_query.rs
Comment (joncinque) [inline cli/src/cluster_query.rs:347]:
```suggestion
                    .long("withdraw-authority"),
```

**Review by joncinque — COMMENTED** [review]
One little comment, otherwise r+ after making the changes Brooks asked for

**Review by brooksprumo — APPROVED** [review]
Looks good to me. I also built and ran this locally and confirmed it worked when providing a withdraw authority. I'll defer to @joncinque or @CriesofCarrots to give another approval before merging. Thanks!

**Comment by brooksprumo** [comment]
I'd also like to have this backported to v1.10. If there are any objections, please remove the v1.10 tag.

**Review by joncinque — APPROVED** [review]
Looks good to me -- thanks for the contribution!

File: cli/src/cluster_query.rs
Comment (CriesofCarrots) [inline cli/src/cluster_query.rs:1809]: This is a bit simpler:
```suggestion
            Some(ref mut filters) => {
                // add filter by withdrawer
                filters.push(withdrawer_filter);
```

Comment (CriesofCarrots) [reply cli/src/cluster_query.rs:1809]: Ah, someone just merged this. Will follow-up myself.

Comment (joncinque) [reply cli/src/cluster_query.rs:1809]: Ah crap, really sorry!

Comment (CriesofCarrots) [reply cli/src/cluster_query.rs:1809]: Nbd! I actually realized we can get rid of the match altogether :) https://github.com/solana-labs/solana/pull/26038
Sadly, `get_or_insert_default` isn't stable yet

**Pull request merged**
